File: src/solver-worker.js
// Web Worker for brute force funding solver calculation

self.addEventListener('message', function(e) {
  const { price022, price023, price024, totalCost, maxQty } = e.data;

  console.log('Worker: Starting calculation...');

  const solutions = [];

  for (let x = 0; x <= maxQty; x++) {
    for (let y = 0; y <= maxQty; y++) {
      for (let z = 0; z <= maxQty; z++) {
        const total = (price022 * x) + (price023 * y) + (price024 * z);
        if (Math.abs(total - totalCost) < 0.01) { // Float comparison with epsilon
          solutions.push([x, y, z]);
        }
      }
    }
  }

  console.log(`Worker: Found ${solutions.length} solutions`);

  if (solutions.length === 0) {
    self.postMessage({ type: 'NoSolutions' });
  } else {
    self.postMessage({ type: 'Solutions', solutions: solutions });
  }
});

File: src/SolverWorker.fs
module FundingSolver.SolverWorker

open Fable.Core
open Fable.Core.JsInterop
open Browser.Types
open FundingSolver.WorkerTypes

// Helper to post message back to main thread
[<Emit("self.postMessage($0)")>]
let postMessage (msg: obj) : unit = jsNative

// Helper to add event listener on self
[<Emit("self.addEventListener($0, $1)")>]
let addEventListener (eventType: string) (handler: MessageEvent -> unit) : unit = jsNative

let solve (data: WorkerMessageData) =
    printfn "Worker: Starting calculation..."
    
    let solutions = ResizeArray<int * int * int>()
    
    for x in 0 .. data.maxQty do
        for y in 0 .. data.maxQty do
            for z in 0 .. data.maxQty do
                let total = (data.price022 * float x) + (data.price023 * float y) + (data.price024 * float z)
                if total = data.totalCost then
                    solutions.Add((x, y, z))
    
    printfn $"Worker: Found {solutions.Count} solutions"
    
    if solutions.Count = 0 then
        postMessage {| ``type`` = "NoSolutions" |}
    else
        postMessage {| ``type`` = "Solutions"; solutions = solutions.ToArray() |}

// Initialize worker - listen for messages
addEventListener "message" (fun (e: MessageEvent) ->
    let data = e.data :?> WorkerMessageData
    solve data
)
